### PR TITLE
NMT bottleneck

### PR DESCRIPTION
**Pull request by michalivne**
This PR adds a bottleneck architecture to NMT models, along with support for training VAE and MIM latent variable models.

# Summary

The bottleneck class (MTBottleneckModel) supports two architectures:
  * No bottleneck: model_type=seq2seq, the usual NMT mdoel.
  * Fixed-size bottleneck: model_type in [seq2seq-br, seq2seq-mim, seq2seq-vae], where the output of the encoder is projected to a fixed number of steps.

The projection to a fixed number of steps is based on the paper <https://arxiv.org/pdf/1703.03130.pdf>.
The idea is to use K attention heads to compute K weighted-average hidden states, projecting a variable number of steps into K steps.

The bottleneck variants offer different losses:
  * The loss of seq2seq-br is reconstruction only (like seq2seq).
  * The loss of seq2seq-mim is reconstruction + latent entropy minimization  (See MIM, <https://arxiv.org/pdf/2003.02645.pdf>).
  * The loss of seq2seq-vae is reconstruction + latent entropy regularization (See <https://arxiv.org/pdf/1312.6114.pdf>) correspondingly.

# YAML Configuration

The following configurations were added to the YAML config:

```
model:
  model_type: 'seq2seq-br' # supports seq2seq, seq2seq-br, seq2seq-mim, seq2seq-vae (see description above)
  min_logv: -8 # minimal allowed logv for seq2seq-mim
  ortho_loss_coef: 0.0 # orthogonality coefficient for attention bridge
  att_bridge_size: 512 # dimension of a step in attention bridge
  att_bridge_k: 16 # fixed number of steps in attention bridge
  att_bridge_inner_size: 1024 # feedforward size in attention bridge
  non_recon_warmup_batches: 200000 # warm-up steps for seq2seq-mim, seq2seq-vae
  recon_per_token: true # when false reconstruction is computed per sample, not per token
```

* ortho_loss_coef - encou
* non_recon_warmup_batches - anneals the KL divergence term (for VAE) or latent entropy term (for MIM)
* recon_per_token - if false, loss is computed per sample (i.e., summed over all tokens), otherwise loss is averaged per token (the default behaviour).

# Usage

See usage example below, training a seq2seq-br with 32 steps.

NOTE: max_generation_delta must be big enough to allow the generation of the longest sequence given the chosen bottleneck size.

```
python -- examples/nlp/machine_translation/enc_dec_nmt-bottleneck.py \
      --config-name=aayn_bottleneck \
      ...
      model.max_generation_delta=256 \
      ...
      model.model_type=seq2seq-br \
      model.att_bridge_size=1024 \
      model.ortho_loss_coef=0.0 \
      model.att_bridge_k=32 \
      model.att_bridge_inner_size=1024 \
      model.recon_per_token=true \
      model.non_recon_warmup_batches=150000 \
      ...
```

# Additional Info

* The attention bridge class (AttentionBridge) was added to ```nemo/collections/nlp/modules/common/transformer/transformer_modules.py```.
* The bottleneck class (MTBottleneckModel) supports logging of various loss terms.

### CODE REVIEW

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Really nice PR! Had a few questions and suggestions :)

File: examples/nlp/machine_translation/conf/aayn_bottleneck.yaml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  preproc_out_dir: null # path to store data preprocessing outputs
+  src_language: 'en'
+  tgt_language: 'de'
+  model_type: 'seq2seq-br' # supports seq2seq, seq2seq-br, seq2seq-mim, seq2seq-vae
+  min_logv: -8 # minimal allowed logv for seq2seq-mim
+  ortho_loss_coef: 0.0 # orthogonality coefficient for attention bridge
+  att_bridge_size: 512 # dimension of a step in attention bridge
+  att_bridge_k: 16 # fixed number of steps in attention bridge
+  att_bridge_inner_size: 1024 # feedforward size in attention bridge
+  non_recon_warmup_batches: 200000 # warm-up steps for seq2seq-mim, seq2seq-vae
```
[the remark this answers is not in the document]
Comment (michalivne) [reply examples/nlp/machine_translation/conf/aayn_bottleneck.yaml:21]: I will make it more descriptive.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py [hunk 2]
```diff
@@ -844,3 +844,383 @@ def list_available_models(cls) -> Optional[Dict[str, str]]:
...
+
+    def forward(self, hidden, hidden_mask=None, return_ortho_loss=False):
+        """
+        Project hidden [B x N x H] to fixed-size [B x k x H]
+
+        return_ortho_loss - if True returns loss term to encourage
+                              orthogonal attention vectors
+        """
+
+        attention_scores = self.W2(self.act(self.W1(hidden) / self.attn_scale) / self.attn_scale).transpose(-1, -2)
```
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:884]: This line only compute the attention score for k attention heads.
The interaction over time is achieved by matrix multiplication the softmax of attention score with the input data (i.e., weighted mean of input data per attention head). Does that make sense?

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py [hunk 3]
```diff
@@ -31,22 +31,22 @@
 from nemo.collections.common.data import ConcatDataset
```
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:31]: I actually head the same thought. Will do.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py [hunk 4]
```diff
@@ -844,3 +844,383 @@ def list_available_models(cls) -> Optional[Dict[str, str]]:
...
+        src_hiddens = self.encoder(
+            input_ids=src,
+            encoder_mask=src_mask,
+        )
+
+        # build posterior distribution q(x|z)
+        z, z_mean, z_logv, bridge_mask, ortho_loss = self.sample_z(
+            hidden=src_hiddens,
+            hidden_mask=src_mask,
+            return_ortho_loss=True,
```
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:1039]: I always return here the otrho_loss and the config defines the weight. Setting it to 0 disables it.
return_ortho_loss is usually False and is required only in forward to allow training using it (hence it is True here and only here).

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py [hunk 5]
```diff
@@ -844,3 +844,383 @@ def list_available_models(cls) -> Optional[Dict[str, str]]:
...
+            encoder_embeddings=bridge_hiddens_dec,
+            encoder_mask=bridge_mask,
+        )
+
+        log_probs = self.log_softmax(hidden_states=tgt_hiddens)
+
+        recon_loss_fn = self.loss_fn if train else self.eval_loss_fn
+
+        if self.recon_per_token:
+            log_p_x_given_z_per_token = -recon_loss_fn(log_probs=log_probs, labels=labels)
```
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:1057]: Since VAE and MIM losses are defined via log probabilities I wanted the code to closely reflect the mathematical formulation. I thought that might be useful for other researchers down the road.
Does that make sense?

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py [hunk 6]
```diff
@@ -844,3 +844,383 @@ def list_available_models(cls) -> Optional[Dict[str, str]]:
...
+            # probability per sample
+            log_p_x_given_z = log_p_x_given_z_per_token.sum(-1).mean()
+
+            tokens = output_mask.sum()
+            log_p_x_given_z_per_token = log_p_x_given_z_per_token.sum().detach() / tokens
+
+        # TODO: replace with a scheduler
+        batch_counter = getattr(self, "batch_counter", 0)
+        if train:
+            self.batch_counter = batch_counter+1
```
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:1079]: Thanks for pointing it out! 
So far I was doing the training in one continuous run so I did not encounter this bug. Will fix.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:31]: Will also move bridge class to collection/nlp module to be able to reuse.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:1039]: Will add comment explaining why return_ortho_loss is True here.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:1057]: I will add all components to be logged in WANDB

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:1079]: Will switch to self.trainer.global_step

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:31]: Moved AttentionBridge to nemo.collections.nlp.modules.common.transformer

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py]: Done for training step. Still need to figure out how to log additional values to test/eval steps.

[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py]: Done for eval step. Will message you when code is tested.

**Comment by ericharper** [comment]
@michalivne, could you add usage instructions to the README?

**Comment by ericharper** [comment]
Please remove the empty file get_wk2.sh

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py [hunk 7]
```diff
@@ -365,6 +365,19 @@ def eval_epoch_end(self, outputs, mode):
...
+                if "log" in x:
+                    for k, v in x["log"].items():
+                        log_dict[k] = log_dict.get(k, []) + [v]
+
+            for k, v in log_dict.items():
+                if dataloader_idx == 0:
+                    self.log(f"{mode}_{k}", np.mean(v), sync_dist=True)
+                else:
+                    self.log(f"{mode}_{k}_dl_index_{dataloader_idx}", np.mean(v), sync_dist=True)
+
```
Comment (ericharper) [inline nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:380]: What's being logged here? when is `"log" in x` ?

Comment (michalivne) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:380]: I can only find NeMo/examples/nlp/language_modeling/get_wkt2.sh which is not empty (and was not added by me). Could you provide a full path for the file?

Comment (MaximumEntropy) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:380]: I think the issue with get_wkt2.sh is not that it is an empty file, but for some reason, the permissions on that file were changed.

Comment (MaximumEntropy) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:380]: From 100644 → 100755. chmod 644 should change it back.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Review by MaximumEntropy — APPROVED** [review]
Thanks for making all of the changes :)
